.storybook/preview.js

```diff
@@ -23,139 +18,215 @@ Object.entries(PrimerBreakpoints).forEach(([viewport, value]) => {
+          const positionB = getPosition(order, valueB)
 
-        // If the stories do not currently match in the same hierarchy, get the
-        // position of each and compare them.
-        const positionA = getPosition(order, valueA)
-        const positionB = getPosition(order, valueB)
+          // If the positions are the same, sort them alphabetically
+          if (positionA === positionB) {
+            return valueA.localeCompare(valueB)
+          }
 
-        // If the positions are the same, sort them alphabetically
-        if (positionA === positionB) {
-          return valueA.localeCompare(valueB)
+          // Otherwise, compare by position in the given order
+          return positionA - positionB
         }
 
-        // Otherwise, compare by position in the given order
-        return positionA - positionB
-      }
-
-      function getHierarchy(story) {
-        const hierarchy = story.title.split('/')
-        hierarchy.push(story.name)
-        return hierarchy
-      }
+        function getHierarchy(story) {
+          const hierarchy = story.title.split('/')
+          hierarchy.push(story.name)
+          return hierarchy
+        }
 
-      return compare(getHierarchy(a), getHierarchy(b))
+        return compare(getHierarchy(a), getHierarchy(b))
+      },
+    },
+    viewport: {
+      viewports: {
+        ...storybookViewports,
+      },
     },
   },
-  viewport: {
-    viewports: {
-      ...storybookViewports,
+}
+
+const primerThemes = [
+  {value: 'light', left: '☀️', title: 'Light'},
+  {value: 'light_colorblind', left: '☀️', title: 'Light Protanopia & Deuteranopia'},
+  {value: 'light_tritanopia', left: '☀️', title: 'Light Tritanopia'},
+  {value: 'light_high_contrast', left: '☀️', title: 'Light High Contrast'},
+  {value: 'dark', left: '🌗', title: 'Dark'},
+  {value: 'dark_dimmed', left: '🌗', title: 'Dark Dimmed'},
+  {value: 'dark_colorblind', left: '🌗', title: 'Dark Protanopia & Deuteranopia'},
+  {value: 'dark_tritanopia', left: '🌗', title: 'Dark Tritanopia'},
+  {value: 'dark_high_contrast', left: '🌗', title: 'Dark High Contrast'},
+]
+
+export const globalTypes = {
+  colorScheme: {
+    name: 'Theme',
+    description: 'Switch themes',
+    defaultValue: 'light',
+    toolbar: {
+      icon: 'contrast',
+      items: [...primerThemes, {value: 'all', left: '', title: 'All'}],
+      showName: true,
+      dynamicTitle: true,
     },
+    showSurroundingElements: {},
   },
 }
+
+export const decorators = [
+  (Story, context) => {
+    document.body.setAttribute('data-color-mode', context.globals.colorScheme.startsWith('light') ? 'light' : 'dark')
+    document.body.setAttribute(
+      'data-light-theme',
+      context.globals.colorScheme.startsWith('light') ? context.globals.colorScheme : undefined,
+    )
+    document.body.setAttribute(
+      'data-dark-theme',
+      context.globals.colorScheme.startsWith('dark') ? context.globals.colorScheme : undefined,
+    )
```

Comment:
I think since decorators are React components we'll want to wrap these in an effect, maybe something like:

```suggestion
    const { colorScheme } = context.globals;
    
    useEffect(() => {
      const colorMode = colorScheme.startsWith('light') ? 'light' : 'dark'
      document.body.setAttribute('data-color-mode', colorMode)
      
      const lightTheme = colorScheme.startsWith('light') ? colorScheme : undefined
      document.body.setAttribute('data-light-theme', lightTheme)
      
      const darkTheme = colorScheme.startsWith('dark') ? colorScheme : undefined
      document.body.setAttribute('data-dark-theme', darkTheme)
    }, [colorScheme]);
   
    
    
```